File: platform/diff-impl/src/com/intellij/openapi/vcs/ex/RangesBuilder.java
/*
 * Copyright 2000-2009 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.intellij.openapi.vcs.ex;

import com.intellij.diff.comparison.ByLine;
import com.intellij.diff.comparison.ComparisonPolicy;
import com.intellij.diff.comparison.DiffTooBigException;
import com.intellij.diff.comparison.TrimUtil;
import com.intellij.diff.comparison.iterables.DiffIterableUtil;
import com.intellij.diff.comparison.iterables.FairDiffIterable;
import com.intellij.diff.util.DiffUtil;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.progress.DumbProgressIndicator;
import com.intellij.openapi.util.Pair;
import com.intellij.openapi.vcs.ex.Range.InnerRange;
import com.intellij.util.diff.FilesTooBigForDiffException;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public class RangesBuilder {
  private static final Logger LOG = Logger.getInstance(RangesBuilder.class);

  @NotNull
  public static List<Range> createRanges(@NotNull Document current, @NotNull Document vcs) throws FilesTooBigForDiffException {
    return createRanges(current, vcs, false);
  }

  @NotNull
  public static List<Range> createRanges(@NotNull Document current, @NotNull Document vcs, boolean innerWhitespaceChanges)
    throws FilesTooBigForDiffException {
    return createRanges(DiffUtil.getLines(current), DiffUtil.getLines(vcs), 0, 0, innerWhitespaceChanges);
  }

  @NotNull
  public static List<Range> createRanges(@NotNull List<String> current,
                                         @NotNull List<String> vcs,
                                         int currentShift,
                                         int vcsShift,
                                         boolean innerWhitespaceChanges) throws FilesTooBigForDiffException {
    try {
      if (innerWhitespaceChanges) {
        return createRangesSmart(current, vcs, currentShift, vcsShift);
      }
      else {
        return createRangesSimple(current, vcs, currentShift, vcsShift);
      }
    }
    catch (DiffTooBigException e) {
      throw new FilesTooBigForDiffException();
    }
  }

  @NotNull
  private static List<Range> createRangesSimple(@NotNull List<String> current,
                                                @NotNull List<String> vcs,
                                                int currentShift,
                                                int vcsShift) throws DiffTooBigException {
    FairDiffIterable iterable = ByLine.compare(vcs, current, ComparisonPolicy.DEFAULT, DumbProgressIndicator.INSTANCE);

    List<Range> result = new ArrayList<>();
    for (com.intellij.diff.util.Range range : iterable.iterateChanges()) {
      int vcsLine1 = vcsShift + range.start1;
      int vcsLine2 = vcsShift + range.end1;
      int currentLine1 = currentShift + range.start2;
      int currentLine2 = currentShift + range.end2;

      result.add(new Range(currentLine1, currentLine2, vcsLine1, vcsLine2));
    }
    return result;
  }

  @NotNull
  private static List<Range> createRangesSmart(@NotNull List<String> current,
                                               @NotNull List<String> vcs,
                                               int currentShift,
                                               int vcsShift) throws DiffTooBigException {
    FairDiffIterable iwIterable = ByLine.compare(vcs, current, ComparisonPolicy.IGNORE_WHITESPACES, DumbProgressIndicator.INSTANCE);

    RangeBuilder rangeBuilder = new RangeBuilder(current, vcs, currentShift, vcsShift);

    for (com.intellij.diff.util.Range range : iwIterable.iterateUnchanged()) {
      int count = range.end1 - range.start1;
      for (int i = 0; i < count; i++) {
        int vcsIndex = range.start1 + i;
        int currentIndex = range.start2 + i;
        if (vcs.get(vcsIndex).equals(current.get(currentIndex))) {
          rangeBuilder.markEqual(vcsIndex, currentIndex);
        }
      }
    }

    return rangeBuilder.finish();
  }

  private static class RangeBuilder extends DiffIterableUtil.ChangeBuilderBase {
    @NotNull private final List<String> myCurrent;
    @NotNull private final List<String> myVcs;
    private final int myCurrentShift;
    private final int myVcsShift;

    @NotNull private final List<Range> myResult = new ArrayList<>();

    public RangeBuilder(@NotNull List<String> current,
                        @NotNull List<String> vcs,
                        int currentShift,
                        int vcsShift) {
      super(vcs.size(), current.size());
      myCurrent = current;
      myVcs = vcs;
      myCurrentShift = currentShift;
      myVcsShift = vcsShift;
    }

    @NotNull
    public List<Range> finish() {
      doFinish();
      return myResult;
    }

    @Override
    protected void addChange(int vcsStart, int currentStart, int vcsEnd, int currentEnd) {
      com.intellij.diff.util.Range range = TrimUtil.expand(myVcs, myCurrent, vcsStart, currentStart, vcsEnd, currentEnd);
      if (range.isEmpty()) return;

      List<InnerRange> innerRanges = calcInnerRanges(range);
      Range newRange = new Range(range.start2, range.end2, range.start1, range.end1, innerRanges);
      newRange.shift(myCurrentShift);
      newRange.vcsShift(myVcsShift);

      myResult.add(newRange);
    }

    @Nullable
    private List<InnerRange> calcInnerRanges(@NotNull com.intellij.diff.util.Range blockRange) {
      try {
        List<String> vcs = myVcs.subList(blockRange.start1, blockRange.end1);
        List<String> current = myCurrent.subList(blockRange.start2, blockRange.end2);

        ArrayList<InnerRange> result = new ArrayList<>();
        FairDiffIterable iwIterable = ByLine.compare(vcs, current, ComparisonPolicy.IGNORE_WHITESPACES, DumbProgressIndicator.INSTANCE);
        for (Pair<com.intellij.diff.util.Range, Boolean> pair : DiffIterableUtil.iterateAll(iwIterable)) {
          com.intellij.diff.util.Range range = pair.first;
          Boolean equals = pair.second;

          byte type = equals ? Range.EQUAL : getChangeType(range.start1, range.end1, range.start2, range.end2);
          result.add(new InnerRange(range.start2 + blockRange.start2, range.end2 + blockRange.start2,
                                    type));
        }
        result.trimToSize();
        return result;
      }
      catch (DiffTooBigException e) {
        return null;
      }
    }
  }

  private static byte getChangeType(int vcsStart, int vcsEnd, int currentStart, int currentEnd) {
    int deleted = vcsEnd - vcsStart;
    int inserted = currentEnd - currentStart;
    if (deleted > 0 && inserted > 0) return Range.MODIFIED;
    if (deleted > 0) return Range.DELETED;
    if (inserted > 0) return Range.INSERTED;
    LOG.error("Unknown change type");
    return Range.EQUAL;
  }
}

File: platform/diff-impl/src/com/intellij/openapi/vcs/ex/RangesBuilder.kt
/*
 * Copyright 2000-2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
@file:JvmName("RangesBuilder")

package com.intellij.openapi.vcs.ex

import com.intellij.diff.comparison.*
import com.intellij.diff.comparison.iterables.DiffIterableUtil
import com.intellij.diff.comparison.iterables.DiffIterableUtil.fair
import com.intellij.diff.comparison.iterables.FairDiffIterable
import com.intellij.diff.tools.util.text.LineOffsets
import com.intellij.diff.tools.util.text.LineOffsetsUtil
import com.intellij.diff.util.DiffRangeUtil
import com.intellij.diff.util.Range
import com.intellij.diff.util.Side
import com.intellij.openapi.editor.Document
import com.intellij.openapi.progress.DumbProgressIndicator
import kotlin.math.max

fun createRanges(current: List<String>,
                 vcs: List<String>,
                 currentShift: Int,
                 vcsShift: Int,
                 innerWhitespaceChanges: Boolean): List<LstRange> {
  val iterable = compareLines(vcs, current)
  return iterable.iterateChanges().map {
    val inner = if (innerWhitespaceChanges) createInnerRanges(vcs.subList(it.start1, it.end1), current.subList(it.start2, it.end2)) else null
    LstRange(it.start2 + currentShift, it.end2 + currentShift, it.start1 + vcsShift, it.end1 + vcsShift, inner)
  }
}

fun createRanges(current: Document, vcs: Document): List<LstRange> {
  return createRanges(current.immutableCharSequence, vcs.immutableCharSequence, current.lineOffsets, vcs.lineOffsets)
}

fun createRanges(current: CharSequence, vcs: CharSequence): List<LstRange> {
  return createRanges(current, vcs, current.lineOffsets, vcs.lineOffsets)
}

private fun createRanges(current: CharSequence,
                         vcs: CharSequence,
                         currentLineOffsets: LineOffsets,
                         vcsLineOffsets: LineOffsets): List<LstRange> {
  val iterable = compareLines(vcs, current, vcsLineOffsets, currentLineOffsets)
  return iterable.iterateChanges().map { LstRange(it.start2, it.end2, it.start1, it.end1) }
}


fun compareLines(text1: CharSequence,
                 text2: CharSequence,
                 lineOffsets1: LineOffsets,
                 lineOffsets2: LineOffsets): FairDiffIterable {
  val range = expand(text1, text2, 0, 0, text1.length, text2.length)
  if (range.isEmpty) return fair(DiffIterableUtil.create(emptyList(), lineOffsets1.lineCount, lineOffsets2.lineCount));

  val contextLines = 5
  val start = max(lineOffsets1.getLineNumber(range.start1) - contextLines, 0)
  val tail = max(lineOffsets1.lineCount - lineOffsets1.getLineNumber(range.end1) - 1 - contextLines, 0)
  val lineRange = Range(start, lineOffsets1.lineCount - tail, start, lineOffsets2.lineCount - tail)

  val iterable = compareLines(lineRange, text1, text2, lineOffsets1, lineOffsets2)
  return fair(DiffIterableUtil.expandedIterable(iterable, start, start, lineOffsets1.lineCount, lineOffsets2.lineCount))
}

fun compareLines(lineRange: Range,
                 text1: CharSequence,
                 text2: CharSequence,
                 lineOffsets1: LineOffsets,
                 lineOffsets2: LineOffsets): FairDiffIterable {
  val lines1 = DiffRangeUtil.getLines(text1, lineOffsets1, lineRange.start1, lineRange.end1)
  val lines2 = DiffRangeUtil.getLines(text2, lineOffsets2, lineRange.start2, lineRange.end2)
  return compareLines(lines1, lines2)
}

fun tryCompareLines(lineRange: Range,
                    text1: CharSequence,
                    text2: CharSequence,
                    lineOffsets1: LineOffsets,
                    lineOffsets2: LineOffsets): FairDiffIterable? {
  val lines1 = DiffRangeUtil.getLines(text1, lineOffsets1, lineRange.start1, lineRange.end1)
  val lines2 = DiffRangeUtil.getLines(text2, lineOffsets2, lineRange.start2, lineRange.end2)
  return tryCompareLines(lines1, lines2)
}

fun fastCompareLines(lineRange: Range,
                     text1: CharSequence,
                     text2: CharSequence,
                     lineOffsets1: LineOffsets,
                     lineOffsets2: LineOffsets): FairDiffIterable {
  val lines1 = DiffRangeUtil.getLines(text1, lineOffsets1, lineRange.start1, lineRange.end1)
  val lines2 = DiffRangeUtil.getLines(text2, lineOffsets2, lineRange.start2, lineRange.end2)
  return fastCompareLines(lines1, lines2)
}


fun createInnerRanges(lineRange: Range,
                      text1: CharSequence,
                      text2: CharSequence,
                      lineOffsets1: LineOffsets,
                      lineOffsets2: LineOffsets): List<LstInnerRange> {
  val lines1 = DiffRangeUtil.getLines(text1, lineOffsets1, lineRange.start1, lineRange.end1)
  val lines2 = DiffRangeUtil.getLines(text2, lineOffsets2, lineRange.start2, lineRange.end2)
  return createInnerRanges(lines1, lines2)
}

private fun compareLines(lines1: List<String>,
                         lines2: List<String>): FairDiffIterable {
  val iwIterable: FairDiffIterable = safeCompareLines(lines1, lines2, ComparisonPolicy.IGNORE_WHITESPACES)
  return processLines(lines1, lines2, iwIterable)
}

private fun tryCompareLines(lines1: List<String>,
                            lines2: List<String>): FairDiffIterable? {
  val iwIterable: FairDiffIterable = tryCompareLines(lines1, lines2, ComparisonPolicy.IGNORE_WHITESPACES) ?: return null
  return processLines(lines1, lines2, iwIterable)
}

private fun fastCompareLines(lines1: List<String>,
                             lines2: List<String>): FairDiffIterable {
  val iwIterable: FairDiffIterable = fastCompareLines(lines1, lines2, ComparisonPolicy.IGNORE_WHITESPACES)
  return processLines(lines1, lines2, iwIterable)
}

/**
 * Compare lines, preferring non-optimal but less confusing results for whitespace-only changed lines
 * Ex: "X\n\nY\nZ" vs " X\n Y\n\n Z" should be a single big change, rather than 2 changes separated by "matched" empty line.
 */
private fun processLines(lines1: List<String>,
                         lines2: List<String>,
                         iwIterable: FairDiffIterable): FairDiffIterable {
  val builder = DiffIterableUtil.ExpandChangeBuilder(lines1, lines2)
  for (range in iwIterable.unchanged()) {
    val count = range.end1 - range.start1
    for (i in 0 until count) {
      val index1 = range.start1 + i
      val index2 = range.start2 + i
      if (lines1[index1] == lines2[index2]) {
        builder.markEqual(index1, index2)
      }
    }
  }

  return fair(builder.finish())
}


private fun createInnerRanges(lines1: List<String>,
                              lines2: List<String>): List<LstInnerRange> {
  val iwIterable: FairDiffIterable = safeCompareLines(lines1, lines2, ComparisonPolicy.IGNORE_WHITESPACES)

  val result = ArrayList<LstInnerRange>()
  for (pair in DiffIterableUtil.iterateAll(iwIterable)) {
    val range = pair.first
    val equals = pair.second
    result.add(LstInnerRange(range.start2, range.end2, getChangeType(range, equals)))
  }
  result.trimToSize()
  return result
}

private fun getChangeType(range: Range, equals: Boolean): Byte {
  if (equals) return LstRange.EQUAL
  val deleted = range.end1 - range.start1
  val inserted = range.end2 - range.start2
  if (deleted > 0 && inserted > 0) return LstRange.MODIFIED
  if (deleted > 0) return LstRange.DELETED
  if (inserted > 0) return LstRange.INSERTED
  return LstRange.EQUAL
}


private fun safeCompareLines(lines1: List<String>, lines2: List<String>, comparisonPolicy: ComparisonPolicy): FairDiffIterable {
  return tryCompareLines(lines1, lines2, comparisonPolicy) ?: fastCompareLines(lines1, lines2, comparisonPolicy)
}

private fun tryCompareLines(lines1: List<String>, lines2: List<String>, comparisonPolicy: ComparisonPolicy): FairDiffIterable? {
  try {
    return ByLine.compare(lines1, lines2, comparisonPolicy, DumbProgressIndicator.INSTANCE)
  }
  catch (e: DiffTooBigException) {
    return null
  }
}

private fun fastCompareLines(lines1: List<String>, lines2: List<String>, comparisonPolicy: ComparisonPolicy): FairDiffIterable {
  val range = expand(lines1, lines2, 0, 0, lines1.size, lines2.size,
                     { line1, line2 -> ComparisonUtil.isEqualTexts(line1, line2, comparisonPolicy) })
  val ranges = if (range.isEmpty) emptyList() else listOf(range)
  return fair(DiffIterableUtil.create(ranges, lines1.size, lines2.size))
}

fun isValidRanges(content1: CharSequence,
                  content2: CharSequence,
                  lineOffsets1: LineOffsets,
                  lineOffsets2: LineOffsets,
                  lineRanges: List<Range>): Boolean {
  val allRangesValid = lineRanges.all {
    isValidLineRange(lineOffsets1, it.start1, it.end1) &&
    isValidLineRange(lineOffsets2, it.start2, it.end2)
  }
  if (!allRangesValid) return false

  val iterable = DiffIterableUtil.create(lineRanges, lineOffsets1.lineCount, lineOffsets2.lineCount)
  for (range in iterable.iterateUnchanged()) {
    val lines1 = DiffRangeUtil.getLines(content1, lineOffsets1, range.start1, range.end1)
    val lines2 = DiffRangeUtil.getLines(content2, lineOffsets2, range.start2, range.end2)
    if (lines1 != lines2) {
      return false
    }
  }
  return true
}

private fun isValidLineRange(lineOffsets: LineOffsets, start: Int, end: Int): Boolean {
  return start in 0..end && end <= lineOffsets.lineCount
}


internal operator fun <T> Side.get(v1: T, v2: T): T = if (isLeft) v1 else v2
internal fun Range.start(side: Side) = side[start1, start2]
internal fun Range.end(side: Side) = side[end1, end2]

internal val Document.lineOffsets: LineOffsets get() = LineOffsetsUtil.create(this)
internal val CharSequence.lineOffsets: LineOffsets get() = LineOffsetsUtil.create(this)